cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll

```diff
@@ -0,0 +1,326 @@
+  IteratorPointerDereferenceOperator() {
+    this.hasName("operator*") and
+    this.getACallToThisFunction().getArgument(0).getFullyConverted().getUnderlyingType() instanceof
+      Iterator
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameter(0) and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * A non-member `operator++` or `operator--` function for an iterator type.
+ */
+class IteratorCrementOperator extends Operator, DataFlowFunction {
+  IteratorCrementOperator() {
+    (
+      this.hasName("operator++") or
+      this.hasName("operator--")
+    ) and
+    this
+        .getACallToThisFunction()
+        .getArgument(0)
+        .getFullyConverted()
+        .getUnderlyingType()
+        .(ReferenceType)
+        .getBaseType() instanceof Iterator
+  }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameter(0) and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * A non-member `operator->` function for an iterator type.
+ */
+class IteratorFieldOperator extends Operator, TaintFunction {
+  IteratorFieldOperator() {
+    this.hasName("operator->") and
+    this
+        .getACallToThisFunction()
+        .getArgument(0)
+        .getFullyConverted()
+        .getUnspecifiedType()
+        .(PointerType)
+        .getBaseType() instanceof Iterator
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameterDeref(0) and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * A non-member `operator+` function for an iterator type.
+ */
+class IteratorAddOperator extends Operator, TaintFunction {
+  IteratorAddOperator() {
+    this.hasName("operator+") and
+    (
+      this
+          .getACallToThisFunction()
+          .getArgument(0)
+          .getFullyConverted()
+          .getUnspecifiedType()
+          .(PointerType)
+          .getBaseType() instanceof Iterator or
+      this
+          .getACallToThisFunction()
+          .getArgument(0)
+          .getFullyConverted()
+          .getUnspecifiedType()
+          .(PointerType)
+          .getBaseType() instanceof Iterator
+    )
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    (
+      input.isParameter(0) or
+      input.isParameter(1)
+    ) and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * A non-member `operator-` function that takes an iterator as its first argument. This includes
+ * both iterator subtraction and iterator difference overloaded operators.
+ */
+class IteratorSubOperator extends Operator, TaintFunction {
+  IteratorSubOperator() {
+    this.hasName("operator-") and
+    this.getACallToThisFunction().getArgument(0).getType().(PointerType).getBaseType() instanceof
+      Iterator
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    (
+      input.isParameter(0) or
+      input.isParameter(1)
```

Comment:
I don't agree that a user-controlled integer `i` transfers taint to `it - i`. Either `i` is out of range, in which case there's UB already at the subtraction, or `i` is in range, in which case `it - i` is tainted iff `it` is tainted.

Same for `operator+`, `operator+=` and `operator-=` in their various member/non-member forms.

For the `operator-` variant that returns an integer, I'd argue that the integer is never tainted -- if there'd been anything wrong with its inputs, like coming from different underlying objects, the subtraction itself would be UB.

Reply:
Adding or subtracting a user-defined integer could be dangerous in itself, but I agree it's probably the job of a query to flag that, rather than the job of taint flow to try and guess the consequences.